### ISSUE TITLE
Intermittent "Peer networkId did not match our own" error

### ISSUE THREAD

**Issue by Nazgolze** [LiskHQ/lisk-sdk#4952]
Originally reported by Nerigal
I reproduced it this morning

### Expected behavior
Lisk Core/SDK is supposed to normal accept connections under normal circumstances that have the correct networkId
### Actual behavior
On occasion, it becomes difficult to or almost impossible to connect to a node without restarting it, with this being the error: `{"event":"#disconnect","data":{"code":4104,"data":"Peer networkId did not match our own"}}`

The networkId is most definitely correct in this case
```
const validateNetworkCompatibility = (peerInfo, nodeInfo) => {
    if (!peerInfo.sharedState) {
        return false;
    }
    if (!peerInfo.sharedState.networkId) {
        return false;
    }
    return peerInfo.sharedState.networkId === nodeInfo.networkId;
};
```
If this is the only place it gets checked, then I suspect it is failing in the first check (in which case, a more correct error message would be helpful)

Here are the logs I obtained when the node had trace turned on (the ip address is one of our seed nodes):
```
lisk.app.log:1843918:2020-03-10 10:22:18 276: {"name":"lisk-framework","module":"network","hostname":"mrbox3","pid":3070,"level":20,"ipAddress":"94.237.44.144","wsPort":5001,"code":4104,"reason":"Peer networkId did not match our own","msg":"EVENT_CLOSE_OUTBOUND: Close outbound peer connection","time":"2020-03-10T10:22:18.276Z","src":{"file":"/home/lisk/lisk-beta/node_modules/lisk-framework/src/modules/network/network.js","line":187},"v":0}
lisk.app.log:1849755:2020-03-10 10:31:38 267: {"name":"lisk-framework","module":"network","hostname":"mrbox3","pid":3070,"level":20,"ipAddress":"94.237.44.144","wsPort":5001,"code":4104,"reason":"Peer networkId did not match our own","msg":"EVENT_CLOSE_OUTBOUND: Close outbound peer connection","time":"2020-03-10T10:31:38.267Z","src":{"file":"/home/lisk/lisk-beta/node_modules/lisk-framework/src/modules/network/network.js","line":187},"v":0}
```

It does, after a while, tend to resolve itself. My theory is because the seed node relearns about my node from other nodes when they exchange peer information, but that's a guess (or it forgets about the peer entirely, that might do it too I think)
### Steps to reproduce
I don't have a way to reproduce this reliably yet, at least not on demand. It happens often enough though. I started a fresh node this morning and turned on trace logging and tcpdump and after about 3 hours or so, it happened

It can be difficult to catch or even notice this because there's little logging here to pin it down exactly, even in trace mode. 
### Which version(s) does this affect? (Environment, OS, etc...)
lisk-core 3.0.0-beta.0
lisk-sdk: 3.0.2 
Linux

**Comment by ishantiw** [comment]
The issue https://github.com/LiskHQ/lisk-sdk/issues/4952 is due to the buggy code we have in lisk-sdk `v3.0.2` version.
https://github.com/LiskHQ/lisk-sdk/blob/0dcfc0389189127066850b6380cf08e28a92f4cf/elements/lisk-p2p/src/p2p.ts#L832-L871
In this section, we check for the incoming peer if it exists or not. if it exists and doesn’t have a `sharedState` or have just `height` 0 then it will not assign incoming `networkId` to it and eventually, it will fail in `validateProtocolVersionCompatibility`
https://github.com/LiskHQ/lisk-sdk/blob/v3.0.2/elements/lisk-p2p/src/utils/validate.ts#L48-L58 and throwing the error as described in the issue.

So the question is, is it possible that a peer can have `sharedState` only having `height: 0`? Yes, it is possible, whenever a peer is discovered through other peers then we can add a peer’s Info with `sharedState` having only `height` equal to zero in the `peerBook` as per the below snippet
https://github.com/LiskHQ/lisk-sdk/blob/0dcfc0389189127066850b6380cf08e28a92f4cf/elements/lisk-p2p/src/utils/sanitize.ts#L54-L77

So what happens is, some peers find you through other peers in the network with only ip and port, and when we add them, we add only height in `sharedState` in the `peerBook`. When we receive an incoming connection from that peer, we check if it's present in the `peerBook`, if yes, then it doesn’t update its `sharedState` that eventually neglects `networkId` coming from query params and it fails in the validation as described above.

This is already being solved in the `development` during refactoring, https://github.com/LiskHQ/lisk-sdk/blob/b9bb1972cb87c6702c41d7b321d154d33831525c/elements/lisk-p2p/src/peer_server/peer_server.ts#L315-L324

**Comment by ishantiw** [comment]
As mentioned, its already solved in #4767
